app/services/storage.js

```diff
@@ -213,11 +222,22 @@ export class ProfileStorage {
 
   async startSession(scope: ?number,
                      ancestor: ?number,
-                     now: number = microtime.now()): Promise<number> {
+                     now: number = microtime.now(),
+                     reason: number = SessionStartReason.newTab): Promise<number> {
+    const result =
+      await this.db
+                .run('INSERT INTO sessionStarts (scope, ancestor, ts, reason) VALUES (?, ?, ?, ?)',
+                     [scope, ancestor, now, reason]);
+    return result.lastID;
+  }
+
+  async endSession(session: number,
+                   now: number = microtime.now(),
+                   reason: number = SessionEndReason.tabClosed): Promise<number> {
```

Comment:
Hmm.  I'd like to consider not typing the reason, at least not at first, to decouple the abstract browsing UA service from the concrete, windows 'n' tabs, UI implementation.  If you're adamant that reasons should be mapped to integers, let's consider having the actual values defined in the UI.

That is, if one was to build an entirely different browser on top of the UA service (a chatbot!), would "tab closed" be a concept?  I think not, it would be something like "chat closed" or "new search started".